www/js/GenWidget.js

```diff
@@ -0,0 +1,158 @@
+/* global Widget, NNE, HTMLElement */
+/*
+  -----------
+     info
+  -----------
+
+  This Code Generator Widget extends the base Widget class to add methods
+  for more easily creating CSS generator Widgets.
+
+  -----------
+     usage
+  -----------
+
+  // this class inherits all the properties/methods of the base Widget class
+  // refer to www/js/Widget.js to see what those are
+  // or take a look at the wiki...
+  // https://github.com/netizenorg/netnet.studio/wiki/Creating-Widgets
+
+  this.codeField = this.createCodeField({
+    value: 'font-size 24px;',
+    change: (e) => this._codeFieldUpdate(e)
+  })
+
+  this.slider = w.createSlider({
+    background: '#f00',
+    min: 0,
+    max: 100,
+    label: 'PX'
+    change: (e) => this._fontSliderUpdate(e)
+  })
+
+*/
+class CodeField extends HTMLElement {
+  connectedCallback (opts) {
+    this.innerHTML = `<div>
+        <button>insert</button>
+        <input type="text">
+    </div>`
+
+    this.input = this.querySelector('input')
+    this.insertButton = this.querySelector('button')
+
+    this.querySelector('button').addEventListener('click', () => {
+      const val = this.querySelector('input').value
+      this.from = NNE.cm.getCursor('from')
+      this.to = NNE.cm.getCursor('to')
+      NNE.cm.replaceSelection(val)
+      const t = { line: this.to.line, ch: this.from.ch + val.length }
+      NNE.cm.setSelection(this.from, t)
+    })
+
+    this.querySelector('input').addEventListener('input', (e) => {
+      this.setAttribute('value', e.target.value)
+      this.change(e)
+    })
+  }
+
+  get value () {
+    return this.getAttribute('value')
+  }
+
+  set value (val) {
+    this.setAttribute('value', val)
+    const input = this.querySelector('input')
+    if (input) {
+      input.value = val
+    }
+  }
+
+  static get observedAttributes () {
+    return ['value']
+  }
+
+  attributeChangedCallback (attrName, oldVal, newVal) {
+    if (newVal !== oldVal) this[attrName] = newVal
+  }
+}
+
+window.customElements.define('code-field', CodeField)
+
+class GenWidget extends Widget {
+  constructor (opts) {
+    super(opts)
+    console.log('genwidget constructed')
+  }
+
+  createCodeField (opts) {
+    const el = document.createElement('code-field')
+    el.setAttribute('value', '12')
+    el.value = 13
+    el.change = opts.change
+
+    return el
+  }
```

Comment:
so the API for this method looks like this:
```js
    this.codeField = this.createCodeField({
      value: 'font-size: 24px;',
      change: (e) => this._codeFieldUpdate(e)
    })
```
u're making use of the `opts.change` callback function correctly, but u're ignoring the `opts.value` that the user passes in. u should be setting the code-field element's value to match what the user passes in. Also, b/c we've created a custom element that has it's attributes && properties in-sync, u don't need both `el.setAttribute('value', opts.value)` && `el.value = opts.value ` here, u only need one or the other (either is fine). 

however, one thing u'll notice after making this update is that the input field will still be blank by default. instead of having it's value set to the value the user passes in. i think this is b/c the assignment (either `el.setAttribute` or `el.value =`) is being called before the `connectedCallback` fires. so u'll also want to update `connectedCallback` so that the input element's initial value is set to `this.getAttribute('value')`